CIPs/0008.md

```diff
@@ -0,0 +1,77 @@
+
+## Goals
+
+- Expand storage and indexing capabilities of this off-chain data
+- Encode the notion of decryption keys
+- Lightweight messaging system
+
+
+## Proposed Solution
+
+We propose to add another claim, called the `STORAGE` claim:
+```json
+{
+  "type": "STORAGE",
+  "timestamp": 1588886114907,
+  "address": "https://example.com/celo_storage",
+  "filterdataPaths": ".*"
+}
+```
+
+This http address should be referred to as the root. The expectation here is that data can accessed via `root/${dataPath}`. A root can specify a REGEX with the `filterdataPaths` key to indicate what kind of data it stores. An account can make multiple `STORAGE` claims, both to increase availability, as well as effectively sharding this off-chain data. This CIP does not specify what to do in case of conflicts.
+
+Here are some common ways of modeling applications into this format:
+
+- For transaction metadata, the account can encrypt the metadata and put it in a file with the transaction hash as the metadata. Further CIPs could specify a standard for thee structure of that metadata.
+    - `dataPath`: `/transactions/transactionHash`
+    - file contents: Encrypted, structured transaction metadata (comments, or otherwise)
+- Data in relation to other accounts could be communicated to other accounts with:
+    - `dataPath`: `/accounts/otherAccountAddress`
+    - file contents: Structured data such as:
+        - `sessionKeys`: The session key(s) encrypted with dataEncryptionKey of `otherAccount` on `Accounts.sol`
+        - `attestations`: These could be signed data attestating things such an audit or badges. An attested account could also store the attestations from the attestor here, as a client can just verify the signature of the attestor.
+- Profile information could be stored on this system with:
+    - `dataPath`: `/profile/name` or `/profile/picture`
+    - file contents: the data unencrypted, or encrypted to session keys available to the recipients
+- Releases:
+    - `dataPath`: `/releases/hashOfRelease`
+    - file contents: The signature of the key on the hash of the release
+
+If no assocation between to the `dataPath` should be leaked, the `dataPath` itself could be encrypted. The recipient would then just have to decrypt all `dataPath`s (which could be available in an index file).
+
+All file contents should always be signed by an authorized signer of the account to prevent tampering by the file hoster. Signatures should be available under `${dataPath}.signature`. It is worth noting that this is only protecting against an adversary like the file hoster trying to provide unauthenticated plaintext. It does not protect against the adversary withholding or returning outdated data. If the order, including the existence, of the data is critical, then the developer should consider falling back to on-chain data, as consensus on the current view of the data is required. A possible partial workaround is to add expiry to the schema.
```

Comment:
In addition to using old data or not returning data, a MitM or the file server may choose to return data for a different path. This could be extremely problematic in cases where, the data from another path is useful in the context of the changed path. For example, the hosting server could return the data for `/accounts/${attackerAddress}/sessionKey` to the query `/accounts/${realAddress}/sessionKey` and the client would be able to verify the signature without knowing the issue.

One solution to this might be to include the path in the signed data

Reply:
That is an extremely good point

Reply:
Added it as a consideration